SingularityService/src/main/java/com/hubspot/singularity/data/NotificationsManager.java

```diff
@@ -0,0 +1,55 @@
+package com.hubspot.singularity.data;
+
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.utils.ZKPaths;
+import org.jetbrains.annotations.NotNull;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.inject.Inject;
+import com.hubspot.singularity.config.SingularityConfiguration;
+
+public class NotificationsManager extends CuratorManager {
+  private static final String NOTIFICATIONS_ROOT = "/notifications";
+  private static final String BLACKLIST_ROOT = NOTIFICATIONS_ROOT + "/blacklist";
+
+  LoadingCache<String, List<String>> cache;
+
+  @Inject
+  public NotificationsManager(CuratorFramework curator, SingularityConfiguration configuration, MetricRegistry metricRegistry) {
+    super(curator, configuration, metricRegistry);
+
+    cache = CacheBuilder.newBuilder()
+        .expireAfterWrite(5, TimeUnit.MINUTES)
+        .build(
+            new CacheLoader<String, List<String>>() {
+              @Override public List<String> load(String key) throws Exception {
+                return getChildren(key);
+              }
+            }
+        );
+  }
+
+  public void addToBlacklist(String email) {
+    create(getEmailPath(email));
+    cache.invalidate(BLACKLIST_ROOT);
```

Comment:
last note here... depending what singularity service instance is hit this may or may not do any good. If the leader is hit with the api call, then it will be effective immediately. If a non-leader instance is hit, the leader will still be waiting until its cache expires. If we want this to be 100% consistent on the leader, we'd want to add the `maybeProxyToLeader` pattern we use in other resources to the NotificationsResource.

Reply:
Good point. But how does hitting the leader ensure synchronization across all instances if each has it's own cache? Wouldn't we need to invalidate the cache on every single instance?

Reply:
It doesn't ensure synchronization everywhere, but it ensures that the one actually doing the email sending is up to date